services/cannon/src/Cannon/API.hs

```diff
@@ -120,24 +132,51 @@ docs (_ ::: url) = do
     return $ responseLBS status200 [jsonContent] doc
 
 push :: UserId ::: ConnId ::: Request -> Cannon Response
-push (user ::: conn ::: req) = do
-    let k = mkKey user conn
+push (user ::: conn ::: req) =
+    singlePush (readBody req) (user, conn) >>= \case
+        PushStatusOk   -> return empty
+        PushStatusGone -> return $ errorRs status410 "general" "client gone"
+
+-- | Parse the entire list of notifcations and targets, then call 'singlePush' on the each of them
+-- in order.
+bulkpush :: Request -> Cannon Response
+bulkpush req =
+    eitherDecodeStrict <$> liftIO (requestBody req) >>= \case
```

Comment:
We use this pattern often: parse request, and throw 400 if it's a bad payload (e.g., https://github.com/wireapp/wire-server/blob/develop/services/brig/src/Brig/API/Handler.hs#L80). Think it'd be good to move this to types-common?

Reply:
will do!

Reply:
d0ecf291b10acdc10bd4064522b11f3d54c6b4cc

Reply:
6cd364a after rebase